cachi2/core/package_managers/pip.py

```diff
@@ -237,6 +257,49 @@ def get_version(self) -> Optional[str]:
         """Attempt to determine the package version. Should only be called if file exists."""
 
 
+class PyProjectTOML(SetupFile):
+    """Extract project name and version from a pyproject.toml file."""
+
+    def __init__(self, top_dir: RootedPath) -> None:
+        """
+        Initialize a PyProjectTOML.
+
+        :param top_dir: Path to root of project directory
+        """
+        super().__init__(top_dir, "pyproject.toml")
+
+    def get_name(self) -> Optional[str]:
+        """Get project name if present."""
+        try:
+            return self._parsed_toml["project"]["name"]
+        except KeyError:
+            log.info("Invalid configuration 'name' is a required property")
+            return None
+
+    def get_version(self) -> Optional[str]:
+        """Get project version if present."""
+        try:
+            return self._parsed_toml["project"]["version"]
+        except KeyError:
+            log.info("Invalid configuration 'version' is a required property")
+            return None
+
+    def check_dynamic_version(self) -> bool:
+        """Check if project version is set dynamically."""
+        try:
+            dynamic_properties = self._parsed_toml["project"]["dynamic"]
+            return "version" in dynamic_properties
+        except KeyError:
+            return False
+
+    @functools.cached_property
+    def _parsed_toml(self) -> dict[str, Any]:
+        with open(self._setup_file, "rb") as f:
+            log.info("Parsing pyproject.toml")
+            data = tomli.load(f)
+            return data
```

Comment:
In case of invalid TOML, we might want to ignore the error rather causing the build to fail (same as setup.cfg and setup.py)